consensus/istanbul/backend/handler.go

```diff
@@ -290,3 +298,177 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// readValidatorProofMessage reads a validator proof message as a part of the
+// istanbul handshake. Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorProofMessage(peer consensus.Peer) (bool, error) {
+	logger := sb.logger.New("func", "readValidatorProofMessage")
+	peerMsg, err := peer.ReadMsg()
+	if err != nil {
+		return false, err
+	}
+
+	var payload []byte
+	if err := peerMsg.Decode(&payload); err != nil {
+		return false, err
+	}
+
+	var msg istanbul.Message
+	err = msg.FromPayload(payload, sb.verifyValidatorProofMessage)
+	if err != nil {
+		return false, err
+	}
+	// If the Msg is empty, the peer has decided not to reveal its info
+	if len(msg.Msg) == 0 {
+		return false, nil
+	}
+
+	var versionedEnode versionedEnode
+	err = rlp.DecodeBytes(msg.Msg, &versionedEnode)
+	if err != nil {
+		return false, err
+	}
+
+	node, err := enode.ParseV4(versionedEnode.Node)
+	if err != nil {
+		return false, err
+	}
+
+	// Ensure the node in the versionedEnode matches the peer node
+	if node.ID() != peer.Node().ID() {
+		logger.Warn("Peer provided incorrect node in versionedEnode", "versionedEnode node", versionedEnode.Node, "peer node", peer.Node().URLv4())
+		return false, errors.New("Incorrect node in versionedEnode")
+	}
+
+	// Check if the peer is within the registered/elected valset
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		logger.Trace("Error in retrieving registered/elected valset", "err", err)
+		return false, err
+	}
+	if !regAndActiveVals[msg.Address] {
+		logger.Debug("Received a validator proof message from peer with address that is not a registered or active validator", "msg.Address", msg.Address)
+		return false, nil
+	}
+
+	// If the versionedEnode message is too old, we don't count this proof as valid
+	// An error is given if the entry doesn't exist, so we ignore the error.
+	knownVersion, err := sb.valEnodeTable.GetVersionFromAddress(msg.Address)
+	if err == nil && versionedEnode.Version < knownVersion {
+		return false, nil
+	}
+
+	// By this point, we know the peer is a validator and we update our val enode table accordingly
+	// Upsert will only use this entry if the version is new
+	err = sb.valEnodeTable.Upsert(map[common.Address]*vet.AddressEntry{msg.Address: {Node: node, Version: versionedEnode.Version}})
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// verifyValidatorProofMessage allows messages that are not signed to still be
+// decoded in case the peer has decided not to identify itself with the validator proof
+// message
+func (sb *Backend) verifyValidatorProofMessage(data []byte, sig []byte) (common.Address, error) {
+	// If the message was not signed, allow it to still be decoded.
+	// A later check will verify if the address is in the validator set, which
+	// will fail.
+	if len(sig) == 0 {
+		return common.ZeroAddress, nil
+	}
```

Comment:
Why not just return an error and fail here? The output of this function could be misinterpreted as signed message from address 0

Reply:
Returning an error and failing will cause the whole handshake to fail, which causes the peer to be dropped.

I'm going to change https://github.com/celo-org/celo-blockchain/pull/873/files#diff-991be8cbcd11f71d672bc2301b763c4cR415 to check if the signature is empty instead of the `Msg` and make it clearer in the comments in the function